<a href="https://colab.research.google.com/github/Jennapa-colab/DWDM21/blob/main/MiniExam.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### ข้อสอบ

จาก data ```flights.csv``` และ ```airports.csv``` ใน https://drive.google.com/drive/folders/1EFo1_uQgWkRiXjezL0T4YwteYrsqU5MV?usp=sharing
0. หารัฐที่มีจำนวนเครื่องบินขาเข้ามากที่สุดและน้อยที่สุด
1. หารัฐที่มีจำนวนเครื่องบินขาออกมากที่สุดและน้อยที่สุด
2. วาด Histogram แสดงจำนวนเครื่องบินขาออกของแต่ละวันในสัปดาห์
3. สายการบินไหนมีเที่ยวบินมากที่สุดและน้อยที่สุด
4. สายการบินอะไรมีความล่าช้ามากที่สุดและน้อยที่สุด
5. ช่วงเช้าหรือช่วงบ่ายมีเครื่องบินขาเข้ามากกว่ากัน

ให้นักศึกษาทำข้อสอบ 2 ข้อจากที่แสดงข้างบนโดยกรอกรหัสนักศึกษาลงใน function ```gen_exams()``` แล้ว run เพื่อให้ได้ข้อสอบที่นักศึกษาต้องทำ 2 ข้อ (ข้อละ 3 คะแนน)
และให้นักศึกษาคิดโจทย์เองอีก 1 ข้อ (4 คะแนน)


In [ ]:
import pandas as pd

In [ ]:
from google.colab import drive 
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
import os

In [ ]:
flights = pd.read_csv('/content/drive/MyDrive/DWDM2021_DATA/flights.csv')
flights

วันในเดือน  วันในสัปดาห์  ... ความล่าช้าขาออก  ความล่าช้าขาเข้า
0                19             5  ...              -3                 1
1                19             5  ...               0                -8
2                19             5  ...              -4               -15
3                19             5  ...              28                24
4                19             5  ...              -6               -11
...             ...           ...  ...             ...               ...
2702213          19             6  ...               1                -3
2702214          19             6  ...              -2                -3
2702215          19             6  ...              18                 9
2702216          19             6  ...              10                 7
2702217          19             6  ...               0               -13

[2702218 rows x 7 columns]

In [ ]:
airports = pd.read_csv('/content/drive/MyDrive/DWDM2021_DATA/airports.csv')
airports

รหัสสนามบิน         เมือง รัฐ                          ชื่อสนามบิน
0          10165   Adak Island  AK                                 Adak
1          10299     Anchorage  AK  Ted Stevens Anchorage International
2          10304         Aniak  AK                        Aniak Airport
3          10754        Barrow  AK      Wiley Post/Will Rogers Memorial
4          10551        Bethel  AK                       Bethel Airport
..           ...           ...  ..                                  ...
360        11233      Cheyenne  WY  Cheyenne Regional/Jerry Olson Field
361        11097          Cody  WY                 Yellowstone Regional
362        11865      Gillette  WY             Gillette Campbell County
363        12441       Jackson  WY                         Jackson Hole
364        14543  Rock Springs  WY       Rock Springs Sweetwater County

[365 rows x 4 columns]

In [ ]:
def gen_exams(ID):
    import math
    a = int(ID[-1])%6
    b = int(ID[-1])%3
    print(f'รหัสนักศึกษา {6230205153} ทำข้อ {a} และ {b if a != b else 5-b}')

ตัวอย่างการ run

In [ ]:
gen_exams('6230205153')

รหัสนักศึกษา 6230205153 ทำข้อ 3 และ 0


In [ ]:
merged_table = flights.merge(airports,how='left',left_on='รหัสสนามบินต้นทาง',right_on='รหัสสนามบิน')
merged_table

วันในเดือน  ...                                 ชื่อสนามบิน
0                19  ...                  Detroit Metro Wayne County
1                19  ...                Salt Lake City International
2                19  ...                      Portland International
3                19  ...             Lambert-St. Louis International
4                19  ...  Cincinnati/Northern Kentucky International
...             ...  ...                                         ...
2702213          19  ...                       Orlando International
2702214          19  ...    Hartsfield-Jackson Atlanta International
2702215          19  ...           Albuquerque International Sunport
2702216          19  ...    Hartsfield-Jackson Atlanta International
2702217          19  ...    Hartsfield-Jackson Atlanta International

[2702218 rows x 11 columns]

In [ ]:
flights.isnull().any() 
#หาค่า missing

วันในเดือน            False
วันในสัปดาห์          False
สายการบิน             False
รหัสสนามบินต้นทาง     False
รหัสสนามบินปลายทาง    False
ความล่าช้าขาออก       False
ความล่าช้าขาเข้า      False
dtype: bool

In [ ]:
airports = pd.read_csv('/content/drive/MyDrive/DWDM2021_DATA/airports.csv')
airports

รหัสสนามบิน         เมือง รัฐ                          ชื่อสนามบิน
0          10165   Adak Island  AK                                 Adak
1          10299     Anchorage  AK  Ted Stevens Anchorage International
2          10304         Aniak  AK                        Aniak Airport
3          10754        Barrow  AK      Wiley Post/Will Rogers Memorial
4          10551        Bethel  AK                       Bethel Airport
..           ...           ...  ..                                  ...
360        11233      Cheyenne  WY  Cheyenne Regional/Jerry Olson Field
361        11097          Cody  WY                 Yellowstone Regional
362        11865      Gillette  WY             Gillette Campbell County
363        12441       Jackson  WY                         Jackson Hole
364        14543  Rock Springs  WY       Rock Springs Sweetwater County

[365 rows x 4 columns]

In [ ]:
airports.isnull().any() 
#หาค่า missing

รหัสสนามบิน    False
เมือง          False
รัฐ            False
ชื่อสนามบิน    False
dtype: bool

##ข้อ 3 สายการบินไหนมีเที่ยวบินมากที่สุดและน้อยที่สุด

In [ ]:
Max = merged_table.sort_values(by=['สายการบิน'],ascending = False)
Max.iloc[:1,[2,3]]

สายการบิน  รหัสสนามบินต้นทาง
1087306        YV              12264

In [ ]:
Min = merged_table.sort_values(by=['สายการบิน'],ascending=True)
Min.iloc[:1,[2,3]]

สายการบิน  รหัสสนามบินต้นทาง
1625949        9E              11433

In [ ]:
MaxMin = flights.groupby('สายการบิน')[['วันในเดือน']].count()
MaxMin

วันในเดือน
สายการบิน            
9E              80031
AA             289855
AS              68555
B6             121906
DL             381657
EV             157928
F9              35738
FL              92702
HA              17432
MQ             113212
OO             160164
UA             286418
US             233321
VX              34739
WN             575739
YV              52821

In [ ]:
MaxMin.reset_index(inplace=True)
MaxMin

สายการบิน  วันในเดือน
0         9E       80031
1         AA      289855
2         AS       68555
3         B6      121906
4         DL      381657
5         EV      157928
6         F9       35738
7         FL       92702
8         HA       17432
9         MQ      113212
10        OO      160164
11        UA      286418
12        US      233321
13        VX       34739
14        WN      575739
15        YV       52821

In [ ]:
MaxMin = MaxMin.sort_values(by=['วันในเดือน'],ascending=False)
MaxMin

สายการบิน  วันในเดือน
14        WN      575739
4         DL      381657
1         AA      289855
11        UA      286418
12        US      233321
10        OO      160164
5         EV      157928
3         B6      121906
9         MQ      113212
7         FL       92702
0         9E       80031
2         AS       68555
15        YV       52821
6         F9       35738
13        VX       34739
8         HA       17432

##สายการบินมีเที่ยวบินมากที่สุด

In [ ]:
Max = MaxMin.sort_values(by=['วันในเดือน'],ascending=False)
Max.iloc[0,:]

สายการบิน         WN
วันในเดือน    575739
Name: 14, dtype: object

##สายการบินที่มีเที่ยวบินน้อยที่สุด

In [ ]:
Min= MaxMin.sort_values(by=['วันในเดือน'],ascending=False)
Min.iloc[-1,:]

สายการบิน        HA
วันในเดือน    17432
Name: 8, dtype: object

##ตอบคือ 

*สายการบินมีเที่ยวบิน คือ สายการบิน WN มีจำนวนเที่ยวบิน 575739 เที่ยวบิน*

*สายการบินมีเที่ยวบิน คือ สายการบิน HA มีจำนวนเที่ยวบิน 17432 เที่ยวบิน*


##ข้อ 0 หารัฐที่มีจำนวนเครื่องบินขาเข้ามากที่สุดและน้อยที่สุด

In [ ]:
merged_table = flights.merge(airports,how='left',left_on='รหัสสนามบินปลายทาง',right_on='รหัสสนามบิน') 
#ตารางซ้ายของตารางหลัก ผูกกับ right_on id
merged_table

วันในเดือน  ...                                ชื่อสนามบิน
0                19  ...                        Miami International
1                19  ...              John F. Kennedy International
2                19  ...               Salt Lake City International
3                19  ...                 Detroit Metro Wayne County
4                19  ...                  Los Angeles International
...             ...  ...                                        ...
2702213          19  ...               Salt Lake City International
2702214          19  ...  Louis Armstrong New Orleans International
2702215          19  ...   Hartsfield-Jackson Atlanta International
2702216          19  ...                      Memphis International
2702217          19  ...          Minneapolis-St Paul International

[2702218 rows x 11 columns]

In [ ]:
merged_table.groupby('รัฐ').count()

วันในเดือน  วันในสัปดาห์  สายการบิน  ...  รหัสสนามบิน   เมือง  ชื่อสนามบิน
รัฐ                                       ...                                  
AK         7700          7700       7700  ...         7700    7700         7700
AZ       100710        100710     100710  ...       100710  100710       100710
CA       380046        380046     380046  ...       380046  380046       380046
CO        97259         97259      97259  ...        97259   97259        97259
CT        13042         13042      13042  ...        13042   13042        13042
DC        80536         80536      80536  ...        80536   80536        80536
FL       208139        208139     208139  ...       208139  208139       208139
GA       148563        148563     148563  ...       148563  148563       148563
HI        28395         28395      28395  ...        28395   28395        28395
IL       176939        176939     176939  ...       176939  176939       176939
IN        18038         18038      18038  ...        18038   18038        18038
KY        10915         10915      10915  ...        10915   10915        10915
LA        25184         25184      25184  ...        25184   25184        25184
MA        66699         66699      66699  ...        66699   66699        66699
MD        51761         51761      51761  ...        51761   51761        51761
MI        62584         62584      62584  ...        62584   62584        62584
MN        60174         60174      60174  ...        60174   60174        60174
MO        63570         63570      63570  ...        63570   63570        63570
NC       104766        104766     104766  ...       104766  104766       104766
NE        13487         13487      13487  ...        13487   13487        13487
NJ        64313         64313      64313  ...        64313   64313        64313
NM        17542         17542      17542  ...        17542   17542        17542
NV        88222         88222      88222  ...        88222   88222        88222
NY       132823        132823     132823  ...       132823  132823       132823
OH        66389         66389      66389  ...        66389   66389        66389
OK        13922         13922      13922  ...        13922   13922        13922
OR        30484         30484      30484  ...        30484   30484        30484
PA        67957         67957      67957  ...        67957   67957        67957
PR        13840         13840      13840  ...        13840   13840        13840
RI         9313          9313       9313  ...         9313    9313         9313
TN        53414         53414      53414  ...        53414   53414        53414
TX       275503        275503     275503  ...       275503  275503       275503
UT        47324         47324      47324  ...        47324   47324        47324
VA        21491         21491      21491  ...        21491   21491        21491
WA        60172         60172      60172  ...        60172   60172        60172
WI        21002         21002      21002  ...        21002   21002        21002

[36 rows x 10 columns]

In [ ]:
merged_table.groupby('รัฐ').count().sort_values(by=['รหัสสนามบินปลายทาง'],ascending=False)

วันในเดือน  วันในสัปดาห์  สายการบิน  ...  รหัสสนามบิน   เมือง  ชื่อสนามบิน
รัฐ                                       ...                                  
CA       380429        380429     380429  ...       380429  380429       380429
TX       275328        275328     275328  ...       275328  275328       275328
FL       208291        208291     208291  ...       208291  208291       208291
IL       176857        176857     176857  ...       176857  176857       176857
GA       148524        148524     148524  ...       148524  148524       148524
NY       132522        132522     132522  ...       132522  132522       132522
NC       104839        104839     104839  ...       104839  104839       104839
AZ       100824        100824     100824  ...       100824  100824       100824
CO        96919         96919      96919  ...        96919   96919        96919
NV        88179         88179      88179  ...        88179   88179        88179
DC        80666         80666      80666  ...        80666   80666        80666
PA        67955         67955      67955  ...        67955   67955        67955
MA        66833         66833      66833  ...        66833   66833        66833
OH        66236         66236      66236  ...        66236   66236        66236
NJ        64071         64071      64071  ...        64071   64071        64071
MO        63593         63593      63593  ...        63593   63593        63593
MI        62728         62728      62728  ...        62728   62728        62728
WA        60237         60237      60237  ...        60237   60237        60237
MN        60015         60015      60015  ...        60015   60015        60015
TN        53470         53470      53470  ...        53470   53470        53470
MD        51679         51679      51679  ...        51679   51679        51679
UT        47410         47410      47410  ...        47410   47410        47410
OR        30579         30579      30579  ...        30579   30579        30579
HI        28402         28402      28402  ...        28402   28402        28402
LA        25189         25189      25189  ...        25189   25189        25189
VA        21539         21539      21539  ...        21539   21539        21539
WI        21005         21005      21005  ...        21005   21005        21005
IN        18085         18085      18085  ...        18085   18085        18085
NM        17558         17558      17558  ...        17558   17558        17558
OK        13902         13902      13902  ...        13902   13902        13902
PR        13787         13787      13787  ...        13787   13787        13787
NE        13516         13516      13516  ...        13516   13516        13516
CT        13071         13071      13071  ...        13071   13071        13071
KY        10945         10945      10945  ...        10945   10945        10945
RI         9324          9324       9324  ...         9324    9324         9324
AK         7711          7711       7711  ...         7711    7711         7711

[36 rows x 10 columns]

In [ ]:
state= merged_table.groupby('รัฐ').count().sort_values(by=['รหัสสนามบินปลายทาง'],ascending=False)
state

วันในเดือน  วันในสัปดาห์  สายการบิน  ...  รหัสสนามบิน   เมือง  ชื่อสนามบิน
รัฐ                                       ...                                  
CA       380429        380429     380429  ...       380429  380429       380429
TX       275328        275328     275328  ...       275328  275328       275328
FL       208291        208291     208291  ...       208291  208291       208291
IL       176857        176857     176857  ...       176857  176857       176857
GA       148524        148524     148524  ...       148524  148524       148524
NY       132522        132522     132522  ...       132522  132522       132522
NC       104839        104839     104839  ...       104839  104839       104839
AZ       100824        100824     100824  ...       100824  100824       100824
CO        96919         96919      96919  ...        96919   96919        96919
NV        88179         88179      88179  ...        88179   88179        88179
DC        80666         80666      80666  ...        80666   80666        80666
PA        67955         67955      67955  ...        67955   67955        67955
MA        66833         66833      66833  ...        66833   66833        66833
OH        66236         66236      66236  ...        66236   66236        66236
NJ        64071         64071      64071  ...        64071   64071        64071
MO        63593         63593      63593  ...        63593   63593        63593
MI        62728         62728      62728  ...        62728   62728        62728
WA        60237         60237      60237  ...        60237   60237        60237
MN        60015         60015      60015  ...        60015   60015        60015
TN        53470         53470      53470  ...        53470   53470        53470
MD        51679         51679      51679  ...        51679   51679        51679
UT        47410         47410      47410  ...        47410   47410        47410
OR        30579         30579      30579  ...        30579   30579        30579
HI        28402         28402      28402  ...        28402   28402        28402
LA        25189         25189      25189  ...        25189   25189        25189
VA        21539         21539      21539  ...        21539   21539        21539
WI        21005         21005      21005  ...        21005   21005        21005
IN        18085         18085      18085  ...        18085   18085        18085
NM        17558         17558      17558  ...        17558   17558        17558
OK        13902         13902      13902  ...        13902   13902        13902
PR        13787         13787      13787  ...        13787   13787        13787
NE        13516         13516      13516  ...        13516   13516        13516
CT        13071         13071      13071  ...        13071   13071        13071
KY        10945         10945      10945  ...        10945   10945        10945
RI         9324          9324       9324  ...         9324    9324         9324
AK         7711          7711       7711  ...         7711    7711         7711

[36 rows x 10 columns]

##รัฐที่มีจำนวนเครื่องบินขาเข้ามากที่สุด

In [ ]:
Max = state1.iloc[:1,5:6]
Max

ความล่าช้าขาออก
รัฐ                 
CA            380429

##รัฐที่มีจำนวนเครื่องบินขาเข้าน้อยที่สุด

In [ ]:
Min = state1.iloc[-1:,5:6] 
Min

ความล่าช้าขาออก
รัฐ                 
AK              7711

##คำตอบ
*รัฐที่มีจำนวนเครื่องบินขาเข้ามากที่สุดคือรัฐ CA มีจำนวนเครื่องบินขาเข้าทั้งหมด 380,429*


รัฐที่มีจำนวนเครื่องบินขาเข้าน้อยที่สุดคือ AK และ มีจำนวนเครื่องบินขาเข้าทั้งหมด 7,711 


##โจทย์คิดเอง รัฐใดที่มีค่าเฉลี่ยของ 'ความล่าช้าขาออก' สูงที่สุด 20 อันดับแรก

In [ ]:
pp = merged_table.groupby('รัฐ').mean().sort_values(by=['ความล่าช้าขาออก'],ascending=False)
pp.iloc[:20,[4]]

ความล่าช้าขาออก
รัฐ                 
NJ         17.528695
NY         13.748630
IL         13.287820
OK         13.260970
VA         12.711593
KY         12.539698
RI         12.215036
CT         12.193482
NE         11.912252
NM         11.620743
MO         11.522888
CA         11.467362
PR         11.300210
CO         11.284588
WI         11.267555
MD         11.001006
OH         10.905142
IN         10.762953
LA         10.543094
PA         10.397307